controllers/storageautoscaler/storageautoscaler_controller.go

```diff
@@ -0,0 +1,255 @@
+package storageautoscaler
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	prometheusapi "github.com/prometheus/client_golang/api"
+	prometheusv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	prometheusconfig "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v4/v1"
+	"github.com/red-hat-storage/ocs-operator/v4/controllers/platform"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// StorageAutoscalerReconciler is the reconciler for StorageAutoscaler objects
+// nolint:revive
+type StorageAutoscalerReconciler struct {
+	client.Client
+	ctx               context.Context
+	Log               logr.Logger
+	Scheme            *runtime.Scheme
+	OperatorNamespace string
+	RequeueTime       time.Duration
+	SyncMap           sync.Map
+	Event             chan event.GenericEvent
+}
+
+// SetupWithManager sets up the reconciler with the manager
+func (r *StorageAutoscalerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// run a go routine to scrape the metrics from prometheus periodically
+	go r.scrapeMetricsPeriodically()
+
+	storageAutoscalerReconcilerController := ctrl.NewControllerManagedBy(mgr).
+		For(&ocsv1.StorageAutoScaler{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		// watch for generic events to trigger the reconcile
+		WatchesRawSource(source.Channel(r.Event,
+			&handler.EnqueueRequestForObject{},
+		))
+
+	return storageAutoscalerReconcilerController.Complete(r)
+}
+
+// +kubebuilder:rbac:groups="monitoring.coreos.com",resources=*,verbs=get
```

Comment:
Using globs (`*`) in resources isn't recommended. Is the plan to set `resources=XYZ` to the minimal list of resources that need perms before merge?

Reply:
Yes, I tried to investigate it but it was time-consuming, we can consider it as a required follow-up. 

In the normal curl operation `prometheus/api` resource worked, but the same was not working with go API, so need to look more onto it

Reply:
You can also try creating a ClusterRoleBinding for this ClusterRole `cluster-monitoring-view` .
Ref: https://github.com/red-hat-storage/ocs-operator/pull/2892

Reply:
Remember the BugZilla issue we had for a long time that several developers spent a long time working on getting rid of `*` in the RBAC items? This is fine for WIP work, but we have already learned that `*` shouldn't be merged into shipped code. This is fine for now, but I'd suggest we resolve this before merging this PR. That or create an ODFBUGS issue for prior to merging this so we are certain it will be fixed before ship.

Reply:
Updated with min required permissions